File: tests/cxx/isce3/geometry/rtc/rtc.cpp
#include <gtest/gtest.h>
#include <isce3/core/Constants.h>
#include <isce3/core/Orbit.h>
#include <isce3/geometry/RTC.h>
#include <isce3/io/IH5.h>
#include <isce3/io/Raster.h>
#include <isce3/product/RadarGridProduct.h>
#include <isce3/product/RadarGridParameters.h>
#include <string>

// Create set of RadarGridParameters to process
std::set<std::string> radar_grid_str_set = {"cropped", "multilooked"};

// Create set of rtcAlgorithms
std::set<isce3::geometry::rtcAlgorithm> rtc_algorithm_set = {
        isce3::geometry::rtcAlgorithm::RTC_BILINEAR_DISTRIBUTION,
        isce3::geometry::rtcAlgorithm::RTC_AREA_PROJECTION};

TEST(TestRTC, RunRTC) {
    // Open HDF5 file and load products
    isce3::io::IH5File file(TESTDATA_DIR "envisat.h5");
    isce3::product::RadarGridProduct product(file);
    char frequency = 'A';

    // Open DEM raster
    isce3::io::Raster dem(TESTDATA_DIR "srtm_cropped.tif");

    // Create radar grid parameter
    isce3::product::RadarGridParameters radar_grid_sl(product, frequency);

    // Crop original radar grid parameter
    isce3::product::RadarGridParameters radar_grid_cropped =
            radar_grid_sl.offsetAndResize(30, 135, 128, 128);

    // Multi-look original radar grid parameter
    int nlooks_az = 5, nlooks_rg = 5;
    isce3::product::RadarGridParameters radar_grid_ml =
            radar_grid_sl.multilook(nlooks_az, nlooks_rg);

    // Create orbit and Doppler LUT
    isce3::core::Orbit orbit = product.metadata().orbit();
    isce3::core::LUT2d<double> dop =
            product.metadata().procInfo().dopplerCentroid(frequency);

    dop.boundsError(false);

    // Set input parameters
    isce3::geometry::rtcInputTerrainRadiometry inputTerrainRadiometry =
            isce3::geometry::rtcInputTerrainRadiometry::BETA_NAUGHT;

    isce3::geometry::rtcOutputTerrainRadiometry outputTerrainRadiometry =
            isce3::geometry::rtcOutputTerrainRadiometry::GAMMA_NAUGHT;

    isce3::geometry::rtcAreaMode rtc_area_mode =
            isce3::geometry::rtcAreaMode::AREA_FACTOR;

    isce3::geometry::rtcAreaBetaMode rtc_area_beta_mode =
            isce3::geometry::rtcAreaBetaMode::AUTO;

    for (auto radar_grid_str : radar_grid_str_set) {

        isce3::product::RadarGridParameters radar_grid;

        // Open DEM raster
        if (radar_grid_str == "cropped")
            radar_grid = radar_grid_cropped;
        else
            radar_grid = radar_grid_ml;

        for (auto rtc_algorithm : rtc_algorithm_set) {

            double geogrid_upsampling = 1;

            std::string filename;
            // test removed because it requires high geogrid upsampling (too
            // slow)
            if (rtc_algorithm == isce3::geometry::rtcAlgorithm::
                                         RTC_BILINEAR_DISTRIBUTION &&
                    radar_grid_str == "cropped") {
                continue;
            } else if (rtc_algorithm == isce3::geometry::rtcAlgorithm::
                                                RTC_BILINEAR_DISTRIBUTION) {
                filename = "./rtc_bilinear_distribution_" + radar_grid_str +
                           ".bin";
            } else {
                filename = "./rtc_area_proj_" + radar_grid_str + ".bin";
            }
            std::cout << "generating file: " << filename << std::endl;

            // Create output raster
            isce3::io::Raster out_raster(filename, radar_grid.width(),
                                        radar_grid.length(), 1, GDT_Float32,
                                        "ENVI");

            // Call RTC
            isce3::geometry::computeRtc(radar_grid, orbit, dop, dem, out_raster,
                    inputTerrainRadiometry, outputTerrainRadiometry,
                    rtc_area_mode, rtc_algorithm, rtc_area_beta_mode,
                    geogrid_upsampling);
        }
    }
}

TEST(TestRTC, CheckResults) {

    for (auto radar_grid_str : radar_grid_str_set) {

        for (auto rtc_algorithm : rtc_algorithm_set) {

            double max_rmse;

            std::string filename;

            // test removed because it requires high geogrid upsampling (too
            // slow)
            if (rtc_algorithm == isce3::geometry::rtcAlgorithm::
                                         RTC_BILINEAR_DISTRIBUTION &&
                    radar_grid_str == "cropped") {
                continue;
            } else if (rtc_algorithm == isce3::geometry::rtcAlgorithm::
                                                RTC_BILINEAR_DISTRIBUTION) {
                max_rmse = 0.7;
                filename = "./rtc_bilinear_distribution_" + radar_grid_str +
                           ".bin";
            } else {
                max_rmse = 0.1;
                filename = "./rtc_area_proj_" + radar_grid_str + ".bin";
            }

            std::cout << "evaluating file: " << filename << std::endl;

            // Open computed integrated-area raster
            isce3::io::Raster testRaster(filename);

            // Open reference raster
            std::string ref_filename =
                    TESTDATA_DIR "rtc/rtc_" + radar_grid_str + ".bin";
            isce3::io::Raster refRaster(ref_filename);
            std::cout << "reference file: " << ref_filename << std::endl;

            ASSERT_TRUE(testRaster.width() == refRaster.width() and
                        testRaster.length() == refRaster.length());

            double square_sum = 0; // sum of square difference
            int n_nan = 0;         // number of NaN pixels
            int n_npos = 0;        // number of non-positive pixels

            // Valarray to hold line of data
            std::valarray<double> test(testRaster.width()),
                    ref(refRaster.width());
            int n_valid = 0;
            for (size_t i = 0; i < refRaster.length(); i++) {
                // Get line of data
                testRaster.getLine(test, i, 1);
                refRaster.getLine(ref, i, 1);
                // Check each value in the line
                for (size_t j = 0; j < refRaster.width(); j++) {
                    if (std::isnan(test[j]) or std::isnan(ref[j])) {
                        n_nan++;
                        continue;
                    }
                    if (ref[j] <= 0 or test[j] <= 0) {
                        n_npos++;
                        continue;
                    }
                    n_valid++;
                    square_sum += pow(test[j] - ref[j], 2);
                }
            }
            printf("    ----------------\n");
            printf("    # total: %d\n", n_valid + n_nan + n_npos);
            printf("    ----------------\n");
            printf("    # valid: %d\n", n_valid);
            printf("    # NaNs: %d\n", n_nan);
            printf("    # non-positive: %d\n", n_npos);
            printf("    ----------------\n");

            ASSERT_GT(n_valid, 0);

            // Compute average over entire image
            double rmse = std::sqrt(square_sum / n_valid);

            printf("    RMSE = %g\n", rmse);
            printf("    ----------------\n");
            
            // Enforce bound on average pixel-error
            ASSERT_LT(rmse, max_rmse);

            // Enforce bound on number of ignored pixels
            ASSERT_LT(n_nan, 1e-4 * refRaster.width() * refRaster.length());
            ASSERT_LT(n_npos, 1e-4 * refRaster.width() * refRaster.length());
        }
    }
}

int main(int argc, char* argv[]) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/python/extensions/pybind/geometry/rtc.py
#!/usr/bin/env python3

import os
import numpy as np
from osgeo import gdal
import isce3.ext.isce3 as isce3
import iscetest
from nisar.products.readers import SLC

# Create list of RadarGridParameters to process
radar_grid_str_list = ['cropped', 'multilooked']

# Create list of rtcAlgorithms
rtc_algorithm_list = [
        isce3.geometry.RtcAlgorithm.RTC_BILINEAR_DISTRIBUTION,
        isce3.geometry.RtcAlgorithm.RTC_AREA_PROJECTION]


def test_rtc():

    # Open HDF5 file and create radar grid parameter
    print('iscetest.data:', iscetest.data)
    h5_path = os.path.join(iscetest.data, 'envisat.h5')
    slc_obj = SLC(hdf5file=h5_path)
    frequency = 'A'
    radar_grid_sl = slc_obj.getRadarGrid(frequency)

    # Open DEM raster
    dem_file = os.path.join(iscetest.data, 'srtm_cropped.tif')
    dem_obj = isce3.io.Raster(dem_file)

    # Crop original radar grid parameter
    radar_grid_cropped = \
            radar_grid_sl.offset_and_resize(30, 135, 128, 128)

    # Multi-look original radar grid parameter
    nlooks_az = 5
    nlooks_rg = 5
    radar_grid_ml = \
            radar_grid_sl.multilook(nlooks_az, nlooks_rg)

    # Create orbit and Doppler LUT
    orbit = slc_obj.getOrbit()
    doppler = slc_obj.getDopplerCentroid()
    doppler.bounds_error = False
    # doppler = isce3.core.LUT2d()

    # set input parameters
    input_terrain_radiometry = isce3.geometry.RtcInputTerrainRadiometry.BETA_NAUGHT
    output_terrain_radiometry = isce3.geometry.RtcOutputTerrainRadiometry.GAMMA_NAUGHT

    rtc_area_mode = isce3.geometry.RtcAreaMode.AREA_FACTOR
    rtc_area_beta_mode = isce3.geometry.RtcAreaBetaMode.AUTO

    for radar_grid_str in radar_grid_str_list:

        # Open DEM raster
        if (radar_grid_str == 'cropped'):
            radar_grid = radar_grid_cropped
        else:
            radar_grid = radar_grid_ml

        for rtc_algorithm in rtc_algorithm_list:

            geogrid_upsampling = 1

            # test removed because it requires high geogrid upsampling (too
            # slow)
            if (rtc_algorithm ==
                        isce3.geometry.RtcAlgorithm.RTC_BILINEAR_DISTRIBUTION and
                radar_grid_str == 'cropped'):
                continue
            elif (rtc_algorithm ==
                       isce3.geometry.RtcAlgorithm.RTC_BILINEAR_DISTRIBUTION):
                filename = './rtc_bilinear_distribution_' + radar_grid_str + '.bin'
            else:
                filename = './rtc_area_proj_' + radar_grid_str + '.bin'

            print('generating file:', filename)

            # Create output raster
            out_raster = isce3.io.Raster(filename, radar_grid.width,
                                         radar_grid.length, 1, gdal.GDT_Float32,
                                         'ENVI')

            # Call RTC
            isce3.geometry.compute_rtc(radar_grid, orbit, doppler, dem_obj, 
                                       out_raster,
                                       input_terrain_radiometry, 
                                       output_terrain_radiometry,
                                       rtc_area_mode, rtc_algorithm,
                                       rtc_area_beta_mode, geogrid_upsampling)

            del out_raster

    # check results
    for radar_grid_str in radar_grid_str_list:
        for rtc_algorithm in rtc_algorithm_list:

            # test removed because it requires high geogrid upsampling (too
            # slow)
            if (rtc_algorithm ==
                        isce3.geometry.RtcAlgorithm.RTC_BILINEAR_DISTRIBUTION and
                radar_grid_str == 'cropped'):
                continue
            elif (rtc_algorithm ==
                       isce3.geometry.RtcAlgorithm.RTC_BILINEAR_DISTRIBUTION):
                max_rmse = 0.7
                filename = './rtc_bilinear_distribution_' + radar_grid_str + '.bin'
            else:
                max_rmse = 0.1
                filename = './rtc_area_proj_' + radar_grid_str + '.bin'

            print('evaluating file:', os.path.abspath(filename))

            # Open computed integrated-area raster
            test_gdal_dataset = gdal.Open(filename)

            # Open reference raster
            ref_filename = os.path.join(
                iscetest.data, 'rtc/rtc_' + radar_grid_str + '.bin')
            
            ref_gdal_dataset = gdal.Open(ref_filename)
            print('reference file:', ref_filename)

            assert(test_gdal_dataset.RasterXSize == ref_gdal_dataset.RasterXSize)
            assert(test_gdal_dataset.RasterYSize == ref_gdal_dataset.RasterYSize)

            square_sum = 0.0 # sum of square difference
            n_nan = 0          # number of NaN pixels
            n_npos = 0          # number of non-positive pixels

            # read test and ref arrays
            test_array = test_gdal_dataset.GetRasterBand(1).ReadAsArray()
            ref_array = ref_gdal_dataset.GetRasterBand(1).ReadAsArray()

            n_valid = 0

            # iterates over rows (i) and columns (j)
            for i in range(ref_gdal_dataset.RasterYSize):
                for j in range(ref_gdal_dataset.RasterXSize):

                    # if nan, increment n_nan
                    if (np.isnan(test_array[i, j]) or np.isnan(ref_array[i, j])): 
                        n_nan = n_nan + 1
                        continue
                    
                    # if n_npos, incremennt n_npos
                    if (ref_array[i, j] <= 0 or test_array[i, j] <= 0):
                        n_npos = n_npos +1
                        continue
                    
                    # otherwise, increment n_valid
                    n_valid = n_valid + 1
                    square_sum += (test_array[i, j] - ref_array[i, j]) ** 2
            print('    ----------------')
            print('    # total:', n_valid + n_nan + n_npos)
            print('    ----------------')
            print('    # valid:', n_valid)
            print('    # NaNs:', n_nan)
            print('    # non-positive:', n_npos)
            print('    ----------------')
            assert(n_valid != 0)

            # Compute average over entire image
            rmse = np.sqrt(square_sum / n_valid)

            print('    RMSE =', rmse)
            print('    ----------------')
            # Enforce bound on average pixel-error
            assert(rmse < max_rmse)

            # Enforce bound on number of ignored pixels
            assert(n_nan < 1e-4 * ref_gdal_dataset.RasterXSize * ref_gdal_dataset.RasterYSize)
            assert(n_npos < 1e-4 * ref_gdal_dataset.RasterXSize * ref_gdal_dataset.RasterYSize)

